<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="configuration_basename" default="turtlebot3_lds_2d.lua"/>
  <arg name="open_rviz" default="true"/>
  <arg name="map_size" default="384"/> <!-- Fixed map size in pixels -->
  <arg name="map_resolution" default="0.05"/> <!-- Map resolution in meters/pixel -->
  
  <!-- Map size calculation -->
  <arg name="map_width_meters" value="$(eval map_size * map_resolution)"/>
  <arg name="map_height_meters" value="$(eval map_size * map_resolution)"/>
  
  <!-- Origin position to center the map -->
  <arg name="origin_x" value="$(eval -map_width_meters / 2)"/>
  <arg name="origin_y" value="$(eval -map_height_meters / 2)"/>
  
  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- SLAM: Gmapping -->
  <node pkg="gmapping" type="slam_gmapping" name="turtlebot3_slam_gmapping" output="screen">
    <param name="base_frame" value="base_footprint"/>
    <param name="odom_frame" value="odom"/>
    <param name="map_frame"  value="map"/>
    
    <!-- Fixed Map Size Parameters -->
    <param name="xmin" value="$(arg origin_x)"/>
    <param name="ymin" value="$(arg origin_y)"/>
    <param name="xmax" value="$(eval map_width_meters / 2)"/>
    <param name="ymax" value="$(eval map_height_meters / 2)"/>
    <param name="delta" value="$(arg map_resolution)"/>
    
    <!-- Map update parameters -->
    <param name="map_update_interval" value="2.0"/>
    
    <!-- Laser parameters -->
    <param name="maxUrange" value="3.0"/>
    <param name="maxRange" value="8.0"/>
    
    <!-- Update/Processing parameters -->
    <param name="sigma" value="0.05"/>
    <param name="kernelSize" value="1"/>
    <param name="lstep" value="0.05"/>
    <param name="astep" value="0.05"/>
    <param name="iterations" value="5"/>
    <param name="lsigma" value="0.075"/>
    <param name="ogain" value="3.0"/>
    <param name="minimumScore" value="0.0"/>
    
    <!-- Motion model parameters -->
    <param name="srr" value="0.1"/>
    <param name="srt" value="0.2"/>
    <param name="str" value="0.1"/>
    <param name="stt" value="0.2"/>
    
    <!-- Likelihood parameters -->
    <param name="linearUpdate" value="1.0"/>
    <param name="angularUpdate" value="0.2"/>
    <param name="temporalUpdate" value="0.5"/>
    <param name="resampleThreshold" value="0.5"/>
    <param name="particles" value="80"/>
    
    <!-- Initial map parameters -->
    <param name="llsamplerange" value="0.01"/>
    <param name="llsamplestep" value="0.01"/>
    <param name="lasamplerange" value="0.005"/>
    <param name="lasamplestep" value="0.005"/>
  </node>

  <!-- RVIZ 
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
          args="-d $(find turtlebot3_slam)/rviz/turtlebot3_gmapping.rviz"/>
  </group> -->
</launch>